### PR TITLE
Add support for `Restart Kernel + Run up to Selected Cell` command

### LINKED ISSUE #9242: Add "Restart and run all" command
### Discussed in https://github.com/microsoft/vscode-jupyter/discussions/9217

<div type='discussions-op-text'>

<sup>Originally posted by **benlindsay** March  3, 2022</sup>
In the Jupyter toolbar, there's a Run All button 

image

which looks a lot like the Restart and Run All button in Jupyter Lab

image

My initial expectation was that those buttons would be analogous, but in VSCode, it does not restart the kernel or clear outputs. Having that button fail to clear outputs is harmful because state from the previous run can be reused. For example, say you have a notebook with 2 cells:

```python
print(foo)
```

and

```python
foo = 42
```

If you run just the second cell, then click "Run All", your notebook will not fail even though it should. I can't think of any reason someone should want to rerun a notebook without also restarting the kernel. As for the clearing outputs bit, there is a case to be made for not clearing outputs, because it can be nice to see what was previously there. However, 2 points in favor of clearing outputs are 1) it would match behavior expectations from Jupyter Lab, and 2) when outputs are cleared prior to rerunning, it's much easier to tell where a failure occurred when scrolling through the notebook. Harder to tell when old outputs remain.

I hacked together a solution by adding a multicommand keybinding to keybindings.json, but that really shouldn't be necessary for a critical functionality. My keybinding looks like this:

```json
{
  "key": "alt+r",
  "command": "extension.multiCommand.execute",
  "args": {
    "sequence": [
      "notebook.clearAllCellsOutputs",
      "jupyter.notebookeditor.restartkernel",
      "notebook.execute"
    ]
  },
  "when": "notebookEditable && activeEditor == 'workbench.editor.notebook'"
}
```

Thanks for all of y'all's work on this project!</div>

### PR DESCRIPTION
**Pull request by Yoyokrazy**
Fixes #9242 

Only thought is a different enablement, taking into account `notebookEditorFocused` and/or `jupyter.hascodecells`? On the other hand, commands like `jupyter.runcurrentcell` only require `isWorkspaceTrusted` to be used. 

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for feature-requests.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/notebooks/notebookCommandListener.ts [hunk 1]
```diff
@@ -175,6 +187,20 @@ export class NotebookCommandListener implements IDataScienceCommandListener {
         this.runAllCells();
     }
 
+    private async RestartKernelAndRunUpToSelectedCell(notebookUri: Uri | undefined) {
```
Comment (rebornix) [inline src/notebooks/notebookCommandListener.ts:190]: restartKernelAndRunUpToSelectedCell

File: src/notebooks/notebookCommandListener.ts [hunk 2]
```diff
@@ -175,6 +187,20 @@ export class NotebookCommandListener implements IDataScienceCommandListener {
...
+        const uri = notebookUri ?? this.notebookEditorProvider.activeNotebookEditor?.notebook.uri;
+
+        await this.restartKernel(uri);
+        if (this.notebooks.activeNotebookEditor) {
```
Comment (rebornix) [inline src/notebooks/notebookCommandListener.ts:194]: Use `uri`.

File: src/notebooks/notebookCommandListener.ts [hunk 3]
```diff
@@ -175,6 +187,20 @@ export class NotebookCommandListener implements IDataScienceCommandListener {
...
+            this.commandManager
```
Comment (DonJayamanne) [inline src/notebooks/notebookCommandListener.ts:195]: I think we should capture this information before line 193 (i.e. before we restart), as its possible for focus to change while we're restarting (sure its unlikely, but who knows what `restartKernel` could do in the future).

**Comment by DonJayamanne** [comment]
Out of curiosity is there any reason why we cannot do something like this https://github.com/microsoft/vscode-jupyter/issues/9242#issuecomment-1489370206

**Comment by Yoyokrazy** [comment]
> Out of curiosity is there any reason why we cannot do something like this [#9242 (comment)](https://github.com/microsoft/vscode-jupyter/issues/9242#issuecomment-1489370206)

Primarily for ease of use and discoverability for the user. It seems like using a dual command as mentioned to bind them together would force the user to search for specifically the term `runCommand`. Ideally there would be a way to add a label, then it would probably be a more desirable solution. Talked with @rebornix and think that core commands that jupyter supports can be implemented as in this PR, and more custom sets of commands can be implemented via `keybindings.json`.

**Review by rebornix — DISMISSED** [review]

File: src/notebooks/notebookCommandListener.ts [hunk 4]
```diff
@@ -175,6 +187,20 @@ export class NotebookCommandListener implements IDataScienceCommandListener {
         this.runAllCells();
     }
 
+    private async restartKernelAndRunUpToSelectedCell(notebookUri: Uri | undefined) {
+        const activeNBE = this.notebookEditorProvider.activeNotebookEditor;
```
Comment (rebornix) [inline src/notebooks/notebookCommandListener.ts:191]: do we need `this.notebookEditorProvider.activeNotebookEditor` here? It's possible that restart happens on an inactive notebook editor but this check will block the restart from running.

File: package.json [hunk 5]
```diff
@@ -626,6 +626,13 @@
                 "category": "Jupyter",
                 "enablement": "isWorkspaceTrusted && (jupyter.interactive.canRestartNotebookKernel || (notebookKernel =~ /^ms-toolsai.jupyter\\// && jupyter.notebookeditor.canrestartNotebookkernel))"
             },
+            {
+                "command": "jupyter.restartkernelandrunuptoselectedcell",
+                "title": "%jupyter.command.jupyter.restartkernelandrunuptoselectedcell.title%",
+                "shortTitle": "%jupyter.command.jupyter.restartkernelandrunuptoselectedcell.title%",
+                "category": "Jupyter",
+                "enablement": "isWorkspaceTrusted && (jupyter.interactive.canRestartNotebookKernel || (notebookKernel =~ /^ms-toolsai.jupyter\\// && jupyter.notebookeditor.canrestartNotebookkernel))"
```
Comment (DonJayamanne) [inline package.json:634]: Wait, why are we adding this new command, i thought this was `Restart and Run all`, this seems to be `Restart and Run upto selected cell`
Why do we need this new command, i didn't realize this new command was a requirement (also i do not think this is something that exists in Jupyter either, hence feels a little odd to add this)

Comment (rebornix) [reply package.json:634]: image

@DonJayamanne it exists in Jupyter Lab.

File: src/notebooks/notebookCommandListener.ts (hunk 4 above)
Comment (Yoyokrazy) [reply src/notebooks/notebookCommandListener.ts:191]: Fair point, then I'll shift to taking the uri as done on L#193 and then it should be covered either case. Just need to change the range on L#197

**Review by rebornix — APPROVED** [review]

**Pull request merged**